employee-rostering-backend/src/main/java/org/optaweb/employeerostering/domain/contract/ContractDTO.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaweb.employeerostering.domain.contract;
+
+public class ContractDTO extends Contract {
```

Comment:
This class is supposed to be used for json serialization - it's used in the controller methods. But it extends the `Contract` class, which contains all the fields and methods, thus this DTO class actually adds no value. It's closely coupled with the `Contract` class by extending it.
That's the opposite to a reason for introducing such a DTO class - to loose the coupling between core domain and json serialization.
The `Contract` class also depends on `ContractDTO`, as it accepts it in one of the constructors. Together with the "extends" relationship, this creates a circular dependency.

Suggestion:
- don't extend `Contract` by the DTO class, rather duplicate the fields of the core domain class and create a standalone DTO, which can be safely used for json serialization
- never create a dependency from core domain class to such a DTO class. The core domain classes represent business rules of your application, while the DTO class is just an implementation detail on how to serialize some data to json. Rather create a constructor here, taking the `Contract` class as a parameter.
- the name `ContractDTO` does not tell much. Neither the location (package) of the class. Does it deal with JPA? Json? Xml? Who knows - before opening the file and reading through it. If you e.g. group a "copy" of the domain classes that is supposed to serve for json serialization into a package that clearly reflects it, you will make it obvious.

Reply:
Very good points, especially on the circular dependency. I plan to keep the DTO objects for reasons I explain below. To fix the circular dependency, I will duplicate the fields from the core domain class and remove the dependent constructor in the `Contract` class. For clarity, I plan to move all the DTO classes to a package called `jsonserialization` so their purpose is clear. Do you agree with this approach?

The reason DTOs were introduced is due to the Spring framework convention. Because Spring MVC automatically binds request parameters to beans declared as arguments of methods annotated with any form of `@RequestMapping` (which is indeed the case for `updatecontract()` and `createContract()`). Passing objects annotated with `@Entity` (`Contract` is an example) directly into controller methods introduces a critical security vulnerability, as users can pass malicious data into the method and directly change unexpected fields in the database. I believe having a DTO class for serialization is necessary due to this reason. It is true that the DTO object adds no business logic, it is purely to mitigate the security flaw, but logic may be added in the future (eg. checking user input).

Reply:
The usage of DTOs is completely fine - not only because of some security flaws, but due to the separation of concerns as I tried to explain => good architecture can actually reduce the risk of security vulnerabilities.
The issue I had with DTOs is that the name does not tell you anything about the purpose - it's very general. Placing them into a package saying the purpose helps. Imagine now, that you have a DTO for service->REST and DTO for database->service. If both classes are named *DTO, it will be hard to distinguish them just by the package. Thus I would actually move them to the right package and also rename them to better reflect their purpose.

Reply:
For inspiration: http://coders-kitchen.com/2013/07/22/serie-choosing-names-1-part-dtos/